### PR TITLE
add markdown support

### PR DESCRIPTION
**Pull request by jorolf**
if you **really** want to emphasize _something_

### CODE REVIEW

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 1]
```diff
@@ -112,16 +115,33 @@ public ChatLine(Message message)
...
+
+            bool bold = false;
+            bool italic = false;
+
+            foreach(string text in message.Content.Split(new string[] { "**" }, StringSplitOptions.None))
+            {
+                string[] textArray = text.Split(new string[] { "*" }, StringSplitOptions.None);
+                for (int i = 0; i < textArray.Length; i++)
+                {
+                    if (i != 0) //we shouldn't switch when the i is zero because then there wasn't a asterisk yet and we don't want to switch at the end because then there could've just been two asterisks
```
Comment (Nabile-Rahmani) [inline osu.Game/Overlays/Chat/ChatLine.cs]: Can't you swap out that toggle for `italic = i % 2 != 0` ?

Comment (Jorolf) [inline osu.Game/Overlays/Chat/ChatLine.cs]: that wouldn't support random bold text in an italic one

**Comment by Tom94** [comment]
First, this needs to support escaping via \. Smilies like *_* are impossible otherwise.
Next, if we go this way, then _ should be permitted for italic text, too, as it is very common in chat applications. Lastly, as far as I can tell the current implementation switches to italic if there is just a single * in a sentence. That should not happen either.

File: osu.Game/Overlays/Chat/ChatLine.cs [hunk 2]
```diff
@@ -112,16 +116,116 @@ public ChatLine(Message message)
...
+            markers.AddRange(parseSplitMarkers(ref toParse, "**", SplitType.Bold));
+            markers.AddRange(parseSplitMarkers(ref toParse, "*", SplitType.Italic));
+            markers.AddRange(parseSplitMarkers(ref toParse, "_", SplitType.Italic));
+
+            // Add a sentinel marker for the end of the string such that the entire string is rendered
+            // without requiring code duplication.
+            markers.Add(new SplitMarker { Index = toParse.Length, Length = 0, Type = SplitType.None });
+
+            // Sort markers from earliest to latest
+            markers.Sort((a, b) => a.Index.CompareTo(b.Index));
```
Comment (huoyaoyuan) [inline osu.Game/Overlays/Chat/ChatLine.cs]: Make `SplitMarker : IComparable<SplitMarker>`, and then default sort method can be used.

Comment (Tom94) [inline osu.Game/Overlays/Chat/ChatLine.cs]: Why overengineer if this is perfectly sufficient? I prefer to leave the struct be just a dumb data container unless the abstractions are actually useful in multiple places.

**Comment by DrabWeb** [comment]
I think this should be moved to the framework with `TextFlowContainer` (that's why it was originally called `Paragraph`, this was planned to be added).

**Comment by peppy** [comment]
I agree we need this in the framework. I'm not sure it makes sense to implement it just for ChatLine here.

**Comment by Tom94** [comment]
If we put it into framework, then it should be a wrapper around the `TextFlowContainer` we have right now. Basically, taking markdown-formatted text and outputting a `TextFlowContainer` which has the correct pieces.

**Comment by peppy** [comment]
Any chance you can reimplement this in the framework as discussed @jorolf? We definitely want to use it in more places than just `ChatLine`.

**Review by peppy — CHANGES_REQUESTED** [review]
Should be added to framework so we can reuse this across multiple use cases.

**Comment by jorolf** [comment]
working on a framework implementation

**Pull request closed without merging**